linkerd/app/inbound/src/direct.rs

```diff
@@ -86,87 +86,79 @@ impl<N> Inbound<N> {
+                            port,
+                            name: Some(name),
+                            protocol,
+                        } => Ok(svc::Either::B(GatewayTransportHeader {
+                            target: NameAddr::from((name, port)),
+                            protocol,
+                            client,
+                        })),
+                        TransportHeader {
+                            name: None,
+                            protocol: Some(_),
+                            ..
+                        } => Err(RefusedNoTarget),
+                    },
+                    // HTTP detection is not necessary in this case, since the transport
+                    // header indicates the connection's HTTP version.
+                    svc::stack(gateway.clone())
+                        .push_on_response(svc::MapTargetLayer::new(io::EitherIo::Left))
+                        .push_map_target(GatewayConnection::TransportHeader)
+                        .instrument(
+                            |g: &GatewayTransportHeader| info_span!("gateway", dst = %g.target),
+                        )
+                        .into_inner(),
+                )
+                // Use ALPN to determine whether a transport header should be read.
+                //
+                // When the transport header is not present, perform HTTP detection to
+                // support legacy gateway clients.
+                .push(NewTransportHeaderServer::layer(detect_timeout))
+                .push_switch(
+                    |client: ClientInfo| {
+                        if client.header_negotiated() {
+                            Ok::<_, Never>(svc::Either::A(client))
+                        } else {
+                            Ok(svc::Either::B(GatewayConnection::Legacy(client)))
+                        }
+                    },
+                    // TODO: Remove this after we have at least one stable release out
+                    // with transport header support.
+                    svc::stack(gateway)
+                        .push_on_response(svc::MapTargetLayer::new(io::EitherIo::Right))
+                        .instrument(|_: &GatewayConnection| info_span!("gateway", legacy = true))
+                        .into_inner(),
+                )
+                .push(rt.metrics.transport.layer_accept())
+                // Build a ClientInfo target for each accepted connection. Refuse the
+                // connection if it doesn't include an mTLS identity.
+                .push_request_filter(ClientInfo::try_from)
+                .push(svc::BoxNewService::layer())
+                .push(tls::NewDetectTls::layer(
+                    rt.identity.clone().map(WithTransportHeaderAlpn),
+                    detect_timeout,
+                ))
+                .check_new_service::<T, I>()
+                .push_on_response(svc::BoxService::layer())
+                .push(svc::BoxNewService::layer())
```

Comment:
since we always add `BoxService` and `BoxNewService` at the end of every stack builder method (and the type signatures require us to do this), I wonder if it makes sense to just make `map_stack` always push `BoxService` and `BoxNewService` methods? that way, it's impossible to forget.

not a blocker, we can add this later.

Reply:
that requires adding type constraints to ensure that stack is a service/newservice, etc. not necessarily a bad idea for a followup, but will make this less flexible (e.g., it couldn't be used for the connect stack).